File: C/doublyLinkedList.c
#include<stdio.h>
#include<stdlib.h>

typedef struct node{
    int data;
    struct node *next;
    struct node *prev;
}node;

typedef struct list{
    node *head;
    node *tail;
}list;

//function declarations
void init(list *l);
void insertBeginning(list *l,int data);
void insertEnd(list *l,int data);
void deleteBeginning(list *l);
void deleteEnd(list *l);
void insertAt(list *l,int data,int pos);
void deleteAt(list *l,int pos);
void reverse(list *l);
void disp(list *l);
int length(list *l);
void clear(list *l);

void init(list *l){
    l->head = NULL;
    l->tail = NULL;
    return;
}

void insertBeginning(list *l, int data){
    node *p;
    p = (node*)malloc(sizeof(node));
    if(p==NULL){
        printf("Inserting unsuccessful\n");
        return;
    }
    p->data = data;
    if(length(l)==0){
        p->next = NULL;
        p->prev = NULL;
        l->head = p;
        l->tail = p;
        printf("%d Inserted successfully\n", data);
        return;
    }
    p->next = l->head;
    l->head->prev = p;
    p->prev = NULL;
    l->head = p;
    printf("%d Inserted successfully\n", data);
    return;
}

void insertEnd(list *l, int data){
    node *p;
    p = (node*)malloc(sizeof(node));
    if(p==NULL){
        printf("Inserting unsuccessful\n");
        return;
    }
    p->data = data;
    if(length(l)==0){
        p->next = NULL;
        p->prev = NULL;
        l->head = p;
        l->tail = p;
        printf("%d inserted successfully\n", data);
        return;
    }
    l->tail->next = p;
    p->prev = l->tail;
    p->next = NULL;
    l->tail = p;
    printf("%d inserted successfully\n", data);
    return;
}

void deleteBeginning(list *l){
    node *p,*q;
    if(length(l)==0){
        printf("List is empty\n");
        return;
    }
    p = l->head;
    if(length(l)==1){
        l->head = NULL;
        l->tail = NULL;
        printf("%d deleted successfully\n", p->data);
        free(p);
        return;
    }
    q=p->next;
    l->head = q;
    q->prev = NULL;
    printf("%d deleted successfully\n",p->data);
    free(p);
    return;
}

void deleteEnd(list *l){
    node *p;
    if(length(l)==0){
        printf("List is empty\n");
        return;
    }
    p = l->tail;
    if(length(l)==1){
        l->head = NULL;
        l->tail = NULL;
        printf("%d deleted successfully\n", p->data);
        free(p);
        return;
    }
    p->prev->next = NULL;
    l->tail = p->prev;
    printf("%d deleted successfully\n", p->data);
    free(p);
    return;
}

void insertAt(list *l, int data, int pos){
    if(pos < 1){
        printf("Invalid position\n");
        return;
    }
    int len = length(l);
    if(pos > len+1){
        printf("Invalid position\n");
        return;
    }
    node *p,*q,*r;
    r = (node*)malloc(sizeof(node));
    if (r == NULL){
        printf("Inserting unsuccessful\n");
        return;
    }
    r->data = data;
    if(pos == 1 && len == 0){
        r->next = l->head;
        r->prev = NULL;
        l->head = r;
        l->tail = r;
        printf("%d inserted successfully\n", data);
        return;
    }
    if(pos == 1){
        r->next = l->head;
        r->prev = NULL;
        l->head = r;
        printf("%d inserted successfully\n", data);
        return;
    }
    if(pos == len+1){
        l->tail->next = r;
        r->prev = l->tail;
        r->next = NULL;
        l->tail = r;
        printf("%d inserted successfully\n", data);
        return;
    }
    p = l->head;
    int i=1;
    while(i!=pos){
        q = p;
        p = p->next;
        i++;
    }
    q->next = r;
    r->prev = q;
    r->next = p;
    p->prev = r;
    printf("%d inserted successfully\n", data);
    return;
}

void deleteAt(list *l, int pos){
    if (pos < 1){
        printf("Invalid position\n");
        return;
    }
    int len = length(l);
    if (len == 0){
        printf("List is empty\n");
        return;
    }
    node *p, *q;
    if (pos > len){
        printf("Invalid position\n");
        return;
    }
    if (pos == 1 && len == 1){
        l->head = NULL;
        l->tail = NULL;
        printf("%d deleted successfully\n",p->data);
        free(p);
        return;
    }
    p = l->head;
    if(pos == 1){
        l->head = p->next;
        l->head->prev = NULL;
        printf("%d deleted successfully\n", p->data);
        free(p);
        return;
    }
    if(pos == len){
        p = l->tail;
        l->tail = p->prev;
        l->tail->next = NULL;
        printf("%d deleted successfully\n", p->data);
        free(p);
        return;
    }
    int i=1;
    p = l->head;
    while (i != pos){
        q = p;
        p = p->next;
        i++;
    }
    q->next = p->next;
    p->next->prev = q;
    printf("%d deleted successfully\n", p->data);
    free(p);
    return;
}

void reverse(list *l){
    if(length(l)==0){
        printf("List is empty\n");
        return;
    }
    if(length(l)==1){
        printf("List reversed\n");
        return;
    }
    node *p,*q,*r;
    r = l->head;
    q = r->next;
    p = q->next;
    r->next = NULL;
    r->prev = q;
    l->tail = r;
    while(p != NULL){
        q->next = r;
        q->prev = p;
        r=q;
        q=p;
        p=p->next;
    }
    q->next = r;
    q->prev = NULL;
    l->head = q;
    printf("List reversed\n");
}

void disp(list *l){
    if(length(l)==0){
        printf("List is empty\n");
        return;
    }
    node *p = l->head;
    printf("List contains: ");
    while(p!=NULL){
        printf("%d ",p->data);
        p = p->next;
    }
    printf("\n");
    return;
}

int length(list *l){
    int length = 0;
    node *p, *q;
    p = l->head;
    while (p != NULL){
        length++;
        p = p->next;
    }
    return length;
}

void clear(list *l){
    node *p,*q;
    p=l->head;
    while(p!=NULL){
        q = p;
        p = p->next;
        free(q);
    }
    l->head = NULL;
    l->tail = NULL;
    printf("List cleared\n");
    return;
}

void main(){
    int loop=1;
    int choice,data,pos;
    list list;
    init(&list);
    printf("\n\t-------------------\n\tDOUBLY LINKED LIST\n\t-------------------\n");
    while(loop){
        printf("\n----------------------------\n1. Insert in the beginning\n2. Insert in the end\n3. Delete from the beginning\n4. Delete from the end\n5. Insert at\n6. Delete at\n7. Reverse the list\n8. Display\n9. Clear\n10. Exit\n");
        printf("\nSelect an option: ");
        scanf("%d",&choice);
        printf("\n");
        switch(choice){
            case 1: //insert beginning
                printf("Enter element: ");
                scanf("%d",&data);
                insertBeginning(&list,data);
                break;

            case 2: //insert end
                printf("Enter element: ");
                scanf("%d", &data);
                insertEnd(&list, data);
                break;
            
            case 3: //delete beginning
                deleteBeginning(&list);
                break;
            
            case 4: //delete end
                deleteEnd(&list);
                break;
            
            case 5: //insert at
                printf("Enter element: ");
                scanf("%d",&data);
                printf("Enter position: ");
                scanf("%d",&pos);
                insertAt(&list,data,pos);
                break;
            
            case 6: //delete at
                printf("Enter position: ");
                scanf("%d", &pos);
                deleteAt(&list,pos);
                break;
            
            case 7: //reverse
                reverse(&list);
                break;
            
            case 8: //display
                disp(&list);
                break;

            case 9: //clear
                clear(&list);
                break;

            case 10: //exit
                clear(&list);
                loop = 0;
                printf("------------EXIT------------\n");
                break;
            
            default:
                printf("Enter a valid option\n");
                break;        
                
        }
    }
}

File: Go/DoublyLinkedList/doublyLinkedList.go
package main

import "fmt"

type Node struct {
	data int
	next *Node
	prev *Node
}

type List struct {
	head *Node
	tail *Node
	length int
}


func initList() *List {
	var list *List = &List{}
	list.head = nil
	list.tail = nil
	list.length = 0
	return list
}


func (list *List) insertBeginning(data int) {
	var newNode *Node = &Node{data: data, next: list.head, prev: nil}

	if(list.length > 0) {
		list.head.prev = newNode
	} 

	list.head = newNode

	if list.length == 0 {
		list.tail = newNode
	}

	list.length += 1
	fmt.Println(data, "Inserted successfully at the beginning")
	return
}


func (list *List) insertEnd(data int) {
	var newNode *Node = &Node{data: data, next: nil, prev: list.tail}

	if(list.length > 0) {
		list.tail.next = newNode
	}

	list.tail = newNode

	if list.length == 0 {
		list.head = newNode
	}

	list.length += 1
	fmt.Println(data, "Inserted successfully at the end")
	return
}


func (list *List) deleteBeginning() {
	if (list.length == 0) {
		fmt.Println("Empty linked list, nothing to delete")
		return
	}

	var p *Node = list.head
	if (list.length == 1) {
		list.head = nil
		list.tail = nil
	} else {
		list.head = list.head.next
		list.head.prev = nil
	}

	list.length -= 1
	fmt.Println(p.data,"Deleted successfully from the beginning")
	p = nil	// marked for the GC
	return 
}


func (list *List) deleteEnd() {
	if(list.length == 0) {
		fmt.Println("Empty linked list, nothing to delete")
		return
	}

	var p *Node = list.tail
	if(list.length == 1) {
		list.head = nil
		list.tail = nil
	} else {
		list.tail = list.tail.prev
		list.tail.next = nil
	}

	list.length -= 1
	fmt.Println(p.data,"Deleted successfully from the end")
	p = nil // marked for the GC
	return
}


func (list *List) insertAt(data int, pos int) {
	if(pos < 0 || pos > list.length) {
		fmt.Println("Invalid position, cannot perform insert")
		return
	}

	if(pos == 0) {
		list.insertBeginning(data)
		return
	}

	if(pos == list.length) {
		list.insertEnd(data)
		return
	}

	var p *Node = list.head
	for i := 0; i < pos-1; i++ {
		p = p.next
	}

	var newNode *Node = &Node{data: data, next: nil, prev: nil}
	newNode.next = p.next
	newNode.prev = p

	p.next = newNode
	newNode.next.prev = newNode

	list.length += 1
	fmt.Println(data, "Inserted successfully")
	return
}


func (list *List) deleteAt(pos int) {
	if(pos < 0 || pos >= list.length) {
		fmt.Println("Invalid position, cannot perform delete")
		return
	}

	if(pos == 0) {
		list.deleteBeginning()
		return
	}

	if(pos == list.length - 1) {
		list.deleteEnd()
		return
	}

	var p *Node = list.head
	for i := 0; i < pos-1; i++ {
		p = p.next
	}

	q := p.next
	p.next = q.next
	q.next.prev = p
	q.next = nil
	q.prev = nil

	list.length -= 1
	fmt.Println(q.data, "Deleted successfully")
	q = nil // marked for the GC
	return
}


func (list *List) reverse() {
	if (list.length == 0) {
		fmt.Println("Empty linked list, nothing to delete")
		return		
	}

	if (list.length == 1) {
		fmt.Print("Reversed | ")
		list.display()
		return
	}

	list.tail = list.head
	r := list.head
	q := r.next
	p := q.next

	r.next = nil
	for (p != nil) {
		q.next = r
		r.prev = q
		r = q
		q = p
		p = p.next
	}

	q.next = r
	r.prev = q
	list.head = q
	fmt.Print("Reversed | ")
	list.display()
	return
}


func (list *List) display() {

	if (list.length == 0) {
		fmt.Println("Empty Linked List! Nothing to display.")
		return
	}

	var p *Node = list.head
	fmt.Print("Linked List: ")
	for (p != nil) {
		fmt.Print(p.data," -> ")
		p = p.next
	}
	fmt.Print("nil | Length: ", list.length,"\n")
	return
}

func (list *List) clear() {
	p := list.head
	q := list.head
	for (p != nil) {
		q = p.next
		p = nil // mark for cleanup by GC
		p = q
	}

	// reset
	list.head = nil
	list.length = 0
	fmt.Println("Linked list cleared")
	return
}

func main() {
	var list *List = initList()
	fmt.Println("Initial Length: ", list.length)

    list.display()
	
	list.insertEnd(3)
	list.insertEnd(4)
	list.insertBeginning(2)
	list.insertBeginning(1)
	
    list.display()
	
	list.deleteEnd()
	list.deleteBeginning()
	list.deleteEnd()
	list.deleteBeginning()

	list.display()

	list.insertEnd(3)
	list.insertEnd(4)
	list.insertBeginning(2)
	list.insertBeginning(1)

	list.insertAt(100, 2)

	list.display()

	list.clear()

	list.insertAt(20, -5)
    list.insertAt(20, 5)
    list.insertAt(20, 0)   // beginning
    list.insertAt(40, 1)   // end
    list.insertAt(80, 2)   // end
    list.insertAt(60, 2)
    list.insertAt(100, 4)  // end
    list.insertAt(50, 2)

    list.display()

    list.deleteAt(-5)
    list.deleteAt(10)
    list.deleteAt(0)    // beginning
    list.deleteAt(4)    // end
    list.deleteAt(2)

    list.display()

    list.reverse()

    list.insertBeginning(1)
    list.insertEnd(100)
    list.insertAt(75, 3)

    list.display()

    list.reverse()

    list.deleteAt(3)

    list.clear()


}